src/codegen/traverse.lisp

```diff
@@ -250,9 +252,47 @@ other nodes, then it would be inappropriate to also define an action
               (fire-action ':traverse (class-name (class-of current-node)) actions args)
               (fire-action ':after    (class-name (class-of current-node)) actions args)
               (fire-action ':after    'node                                actions args))))
-      (let ((*traverse* #'current-traverse))
+      (let ((*traverse* (lambda (node &rest args)
+                          (funcall *traverse-hook*
+                                   (lambda ()
+                                     (apply #'current-traverse node args))
+                                   node))))
         (apply *traverse* initial-node initial-args)))))
 
+(defun identity-hook (thunk node)
+  "Default traversal hook function, which doesn't do any extra operation."
+  (declare (ignore node)) (funcall thunk))
+
+(defvar *traverse-hook* #'identity-hook)
+
+(defun traverse-with-hook (hook node action-list &rest args)
+  "Like `traverse`, but uses `hook` as the traversal hook function.
+The hook function is called whenever `traverse` recurse into the sub
+nodes.  It takes two arguments: `thunk` to recurse traversal, and `node`
+that is about to be recursed.
+
+One of the use cases is to record parent node of the node being traversed.
+See `print-node-parent` in traverse.lisp for the example."
+  (let ((*traverse-hook* hook))
+    (apply #'traverse node action-list args)))
+
+;; Used in traverse-with-parent-ndoe
+(defvar *traverse-parent-node-parent* nil)
+(defvar *traverse-parent-node-current* nil)
+
+(defun traverse-with-parent-thunk (node action-list &rest args)
+  "Like 'traverse', but actions are called with a thunk in the first argument,
+which returns the node's parent node.
+This captures one of the typical use case of `traverse-with-hook`."
+  (labels ((track-parent-node-hook (thunk node)
+             (let ((*traverse-parent-node-parent* *traverse-parent-node-current*)
+                   (*traverse-parent-node-current* node))
+               (funcall thunk))))
+    (let ((*traverse-hook* #'track-parent-node-hook))
```

Comment:
traverse-with-hook can be used here, no?

Reply:
Right.  That was the original intention but somehow I forgot, I guess.